### ISSUE TITLE
Very fast convergence (and overfitting?), and useless PostNet

### ISSUE THREAD

**Issue by alexdemartos** [as-ideas/ForwardTacotron#12]
Hi again,

I would like to ask for your experience with this model. In my case, I found the model converges very fast, even with 10K steps training (batch size 64). I tried significantly reducing model size (from 20M to 10M params), but no difference. I am training on a subset of VCTK multi-speaker corpus. After that, the validation loss keeps increasing (even though I see no degradation on inference quality).

![tensorboard](https://user-images.githubusercontent.com/2833410/84998681-6fd5bd00-b150-11ea-8a14-7560b70879e9.png)

Another thing that I noticed is that the Postnet seems to be doing nothing relevant (before and after means b/a the Postnet). Is that also your case?

Thanks for your advice.

**Comment by alexdemartos** [comment]
My intuition on the PostNet providing no improvements is that, in the case of Tacotron, the PostNet is used because the decoder sees no information from future time-steps (it is an uni-directional decoding). In this model, the decoder has all the alignment information beforehand, and the decoder is able to run in both directions, being then powerful enough to reconstruct the mel spectrogram with no help of an additional network.

**Comment by alexdemartos** [comment]
Regarding the early overfitting, I wonder if this is related to removing the Prenet bottleneck for the input embeddings (2 FC + 0.5 dropout) before feeding them to the encoder/CBHG.

Quote from original Tacotron paper:
_We use a bottleneck layer with dropout as the pre-net in this work, which helps convergence and improves generalization._

Did you remove it for a specific reason?

Thanks again, will keep you posted as I try different solutions :)

**Comment by cschaefer26** [comment]
Hi, I also found the val loss going up quickly. Still the sound improves up until 180k steps or so, my reasoning is that the produced mel patterns improve but differ from the ground truth as the model is not teacher forced as the original tacotron. As for the postnet, I have seen quite some difference between linear and postnet output, I think here i simply adds some model parameters than improve mel quality, nothing special.

**Comment by alexdemartos** [comment]
That's true, original Tacotron's validation loss only measures how well the decoder is able to predict next mel based on the (target) previous one. I guess this is not useful when there's no auto-regression and the model is free to "pronounce differently" the validation sentences.

Nevertheless, I found adding a bottleneck PreNet to the input embeddings indeed helps generalization. The overfitting seems to be less pronounced, while the training loss curves seem more stable as well.

![Captura](https://user-images.githubusercontent.com/2833410/85052279-ec8c8980-b198-11ea-8d92-a0a5936008a7.JPG)

**Comment by cschaefer26** [comment]
Cool, thanks for sharing! I gonna do another run with the prenet and also some other stuff (I can only do one test at a time with a single GPU):

- Add encoder PreNet
- Duration predictor on top of encoder
- L2 distance on log durations

I am however hesitant to remove the postnet, see the images below:

before postnet
![before](https://user-images.githubusercontent.com/54799962/85101288-59883980-b202-11ea-8028-ddc57b5e5ba6.png)

after postnet
![after](https://user-images.githubusercontent.com/54799962/85101293-5c832a00-b202-11ea-915a-a1af2954fa1b.png)

**Comment by m-toman** [comment]
I'm seeing a small difference in loss with/without postnet
![image](https://user-images.githubusercontent.com/9433378/85221093-626b3d80-b3b1-11ea-9d7f-4677fcba6aa6.png)

Blue one was the original run with post-net - mel (L1) loss of the postnet.
Red one was where the same loss was just a copy of m1

(so blue:
        x_post = self.postnet(x)
        x_post = self.post_proj(x_post)
        x_post = x_post.transpose(1, 2)
red:
        x_post = x
)

**Comment by alexdemartos** [comment]
In both your cases the Postnet seems to be doing a significant job. In my case (where I am using Taco2 residual Postnet instead of the CBHG Postnet proposed in this repo) it seems to provide no relevant enhancement on the reconstruction of the mel spectra.

**Comment by m-toman** [comment]
I've also added the prenet after embedding and before cbhg now, training.
But as I see it, we always get the same embedding per input symbol of dim 512. Then apply the same two 256 dim FCs without any convolution or RNN before it (vs the prenet in the autoregressive loop) so still end up with the same 256 dim vector per input symbol. 
Wonder if there's much to gain there?

**Comment by alexdemartos** [comment]
Hi Markus,

I think the key is in the high dropout applied in the PreNet, which in my opinion helps generalization and reduces overfitting. I also predict durations on the outputs of the CBHG instead of on the input embeddings, which also seems to help prosody.

**Comment by m-toman** [comment]
Yeah I also changed that following your suggestion.
I temoved this dropout with the introduction of the prenet
https://github.com/as-ideas/ForwardTacotron/blob/c2e03f1ad4ddcdb3efc58118dfd51b5cc8cca53e/models/forward_tacotron.py#L166 

No experiment on it but thought may be a bit too much with both of them. 
Taco2 also uses dropout at inference time. Wonder if we need that too.. I would suspect not.

**Comment by cschaefer26** [comment]
I think the dropout in inference helps with attention being stuck at pauses, in the taco implementation of fatchord its only used in training.
